com.unity.render-pipelines.universal/Runtime/LightCookieManager.cs

```diff
@@ -115,6 +120,97 @@ public void Resize(int size)
             }
         }
 
+        private struct ShaderBitArray
+        {
+            const int kBitsPerElement = 32;
+            const int kElementShift = 5;
+            const int kElementMask = (1 << kElementShift) - 1;
+
+            private float[] m_data;
+
+            public int elemLength => m_data == null ? 0 : m_data.Length;
+            public int bitCapacity => elemLength * kBitsPerElement;
+            public float[] data => m_data;
+
+            public void Resize(int bitCount)
+            {
+                if (bitCapacity > bitCount)
+                    return;
+
+                int newElemCount = ((bitCount + (kBitsPerElement - 1)) / kBitsPerElement);
+                if (newElemCount == m_data?.Length)
+                    return;
+
+                var newData = new float[newElemCount];
+                if (m_data != null)
+                {
+                    for (int i = 0; i < m_data.Length; i++)
+                        newData[i] = m_data[i];
+                }
+                m_data = newData;
+            }
+
+            public void Clear()
+            {
+                for (int i = 0; i < m_data.Length; i++)
+                    m_data[i] = 0;
+            }
+
+            private void GetElementIndexAndBitOffset(int index, out int elemIndex, out int bitOffset)
+            {
+                elemIndex = index >> kElementShift;
+                bitOffset = index & kElementMask;
+            }
+
+            public bool this[int index]
+            {
+                get
+                {
+                    GetElementIndexAndBitOffset(index, out var elemIndex, out var bitOffset);
+
+                    unsafe
+                    {
+                        fixed(float* floatData = m_data)
+                        {
+                            uint* uintElem = (uint*)&floatData[elemIndex];
+                            bool val = ((*uintElem) & (1u << bitOffset)) != 0u;
+                            return val;
+                        }
+                    }
+                }
+                set
+                {
+                    GetElementIndexAndBitOffset(index, out var elemIndex, out var bitOffset);
+                    unsafe
+                    {
+                        fixed(float* floatData = m_data)
+                        {
+                            uint* uintElem = (uint*)&floatData[elemIndex];
+                            if (value == true)
+                                *uintElem = (*uintElem) | (1u << bitOffset);
+                            else
+                                *uintElem = (*uintElem) & ~(1u << bitOffset);
+                        }
+                    }
+                }
+            }
+
+            public override string ToString()
+            {
+                unsafe
+                {
+                    int len = bitCapacity;
+                    byte* buf = stackalloc byte[len];
```

Comment:
We do not have a cap on bitCapacity. As I recall it is strongly recommended to have a hard limit when using stackalloc and have a fallback path using non stackalloc array when this limit is exceeded.

Reply:
Technically we do, maxLights / 32, but yes it would be better to not trust that as that can change.

Reply:
I decided to just clamp it just in case. It's a private method and mostly for debug.